Prevent infinite resize when vertical scrollbar appears

kurkle opened the pull request:
A workaround to the _infinte resize_ problem.

Pens to demonstrate:  
[master](https://codepen.io/kurkle/pen/YBXpxW)
[master - 3 columns](https://codepen.io/kurkle/pen/YBwBjK)

THIS PR:(https://github.com/chartjs/Chart.js/pull/6011/commits/840d11a8d32260849f9ff3ad0b4f775eda99e9c2)
[single](https://codepen.io/kurkle/pen/XObNgQ) 
[3 rows](https://codepen.io/kurkle/pen/QYyNRQ)
[3 columns](https://codepen.io/kurkle/pen/qgbNbd)
[collapse rows](https://codepen.io/kurkle/pen/xMZOrw)
[collapse columns](https://codepen.io/kurkle/pen/qgbNvr)

You get the effect (in master) by resizing window so that scrollbar just appears.

Fixes: #2127

kurkle commented on src/core/core.controller.js:
```diff
@@ -213,13 +213,13 @@ helpers.extend(Chart.prototype, /** @lends Chart */ {
 			plugins.notify(me, 'resize', [newSize]);
 
 			// Notify of resize
-			if (me.options.onResize) {
```
options is already cached, remove unnessesary me.

etimberg commented on src/platforms/platform.dom.js:
```diff
@@ -277,13 +277,10 @@ function addResizeListener(node, listener, chart) {
+			if (container && container.clientWidth !== w && chart.canvas) {
+				// If the container size changed during chart resize, we can assume scrollbar appeared.
+				// So let's resize again, with the scrollbar visible
+				ret = listener(createEvent('resize', chart));
```
I'm curious how the 2nd resize doesn't cause the scrollbar to go away

kurkle replied:
What _I think_ happens: We are doing the resize in a `onScroll` event. We resize, notice it made scrollbars visible and resize again. Scrollbars are hidden in the end of this scroll event. And I suppose browsers won't fire another scroll until the previous is done.

Initially I was string the width, but accidentaly found out, it was not really needed.

etimberg replied:
Ahh, that's a good find. If `throttled()` can prevent the double event then I'm ok with testing this out further

simonbrunel commented on src/platforms/platform.dom.js:
```diff
@@ -268,12 +268,21 @@ function unwatchForRender(node) {
 	var resizer = expando.resizer = createResizer(throttled(function() {
 		if (expando.resizer) {
-			return listener(createEvent('resize', chart));
+			var aspectRatio = chart.options.maintainAspectRatio && chart.aspectRatio || null;
```
`var aspectRatio = chart.options.maintainAspectRatio;` should be enough

kurkle replied:
To be consistent, I copied this from `resize`: https://github.com/chartjs/Chart.js/blob/e6a0c8690981de60d789f74e605f6d6b4557e3f6/src/core/core.controller.js#L190

simonbrunel commented on src/platforms/platform.dom.js:
```diff
@@ -268,12 +268,21 @@ function unwatchForRender(node) {
+			var container = aspectRatio && node.parentNode;
+			var w = container ? container.clientWidth : 0;
+			var ret = listener(createEvent('resize', chart));
+			if (container && container.clientWidth !== w && chart.canvas) {
```
I don't think we need to test `chart.canvas`

kurkle replied:
Some tests `destroy` chart after the first resize event (`canvas = null` there). So those throw without this test. Originally I placed this check in `resize`, but moved it here instead.

kurkle commented on src/platforms/platform.dom.js:
```diff
@@ -268,12 +268,21 @@ function unwatchForRender(node) {
 
 function addResizeListener(node, listener, chart) {
-	var expando = node[EXPANDO_KEY] || (node[EXPANDO_KEY] = {});
+	var expando = node[EXPANDO_KEY] = node[EXPANDO_KEY] || {};
```
I see your point and there is a clear performance difference: 
https://jsperf.com/assign-if-empty-always-vs-conditional
However, this is called only once per chart, so I thought minification might be more important.
I'll revert this.

simonbrunel replied:
> I thought minification might be more important

I think performance (and logic) is always more important than minification, especially when the difference is 2 bytes :)

```
node[EXPANDO_KEY]||(node[EXPANDO_KEY]={})
node[EXPANDO_KEY]=node[EXPANDO_KEY]||{}
```

kurkle commented on src/platforms/platform.dom.js:
```diff
@@ -268,12 +268,21 @@ function unwatchForRender(node) {
+			var aspectRatio = chart.options.maintainAspectRatio && chart.aspectRatio || null;
+			var container = aspectRatio && node.parentNode;
+			var w = container ? container.clientWidth : 0;
+			var ret = listener(createEvent('resize', chart));
```
Old was `return listener...` and I was not completely sure if there was some sort of cancellation possibility there, so I stored it. But if its not needed, then I'll remove it.

simonbrunel replied:
I think `listener` returns nothing and we don't use the [`handler()` result](https://github.com/chartjs/Chart.js/blob/master/src/platforms/platform.dom.js#L223).

kurkle commented:
> I'm still getting a jumpy behavior (not infinite) just before the scrollbar become visible or hidden. I think it's better than master but still unexpected and will certainly be reported as a bug.

This is due to the 2nd resize that's done while scrollbar is visible (and thus reducing width).
The width is a bit smaller (scrollbar width -1 or so) than just before, so it jumps a little.
I thought about adding the math to compensate that part, but I think its not worth the effort and prone to buggy situations.

> Anyway, I'm a bit worry about this PR because we are considering that a container width change during resize is due to scrollbar visibility change. But what about animations (via JS or CSS) or similar? can you setup a pen that resize the container with a CSS animation (expand / collapse for example)?

Good points, I'll do some more pens. 

> Also, can we test this behavior with multiple charts on the same page (same row or not)?

I've tested it with multiple charts on different lines and they end up being different sizes some times. Still better than infinite resize IMO. I'll add pens for this too.

simonbrunel commented:
> I thought about adding the math to compensate that part, but I think its not worth the effort and prone to buggy situations.

Agreed, and again, we can't assume that the container width has changed because the scrollbar visibility change. Also, we would need to handle cases where the scrollbars are not in the body but on any parent container.

> I've tested it with multiple charts on different lines and they end up being different sizes some times. Still better than infinite resize IMO.

"Better" if that happens instead of the infinite resize only. What I'm worried about is that it could happen in other scenarios than scrollbars.

simonbrunel commented:
Thanks @kurkle for the pens, I can't barely test the `columns` ones because my screen is not height enough (laptop) :). `3 columns` jumps a lot vertically when resizing, I'm wondering if it's the same issue with master? Else size animations don't seem to be impacted by this workaround.

kurkle commented:
@simonbrunel change the aspectRatio to suit your screen better :)

simonbrunel commented:
Yeah, that's what I did but can't test the pen in "Full page view", but it was enough to see the "obvious issue here". Having the full Chart.js code in the pen makes the pen really painful to manipulate, I'm wondering if there is not another way to easily integrate WIP builds.

simonbrunel commented on src/platforms/platform.dom.js:
```diff
@@ -273,7 +273,14 @@ function addResizeListener(node, listener, chart) {
+			var w = container ? container.clientWidth : 0;
+			listener(createEvent('resize', chart));
+			if (container && container.clientWidth !== w && chart.canvas) {
+				// If the container size changed during chart resize, we can assume scrollbar appeared.
```
nit: I would break the comment after "assume"

kurkle commented:
Limited the 2nd resize to when container size shrinks in the first one. This removes some jumpiness (and is logical).

Updated the comment to contain all relevant information (I'm not native in English, so happy to reword).

All pens updated and also added a `master - 3 columns`.
In the ones with columns, I added `vertical-align: top;` to css.

ashley-mort commented:
> I'm still getting a jumpy behavior (not infinite) just before the scrollbar become visible or hidden. I think it's better than master but still unexpected and will certainly be reported as a bug.
> 
> Anyway, I'm a bit worried about this PR because we are considering that a container width change during resize is due to scrollbar visibility change. But what about animations (via JS or CSS) or similar? can you setup a pen that resize the container with a CSS animation (expand / collapse for example)?
> 
> Also, can we test this behavior with multiple charts on the same page (same row or not)?

I still see the infinite scroll even just before my Y axis scrollbar appears.  Any tips on how to avoid this?  Chart.js 2.9.4
